### DISCUSSION TITLE
It was tough to add middleware

### DISCUSSION THREAD

**Discussion by kulak** [tokio-rs/axum#2222] [General]
While I like the general design `axum` takes there is extra cost to be paid compared to other languages. This is story of how tough it was.

1st I tried a standalone layer middleware and the most difficult part was figuring out how to process async validation request and pass it to next service only on success.  This was exercise in figuring out how to run async validation task without building next Future.  In the end a custom ResponseFuture with cloned `inner` or `next` execution step is what axum expects to do.

I found an example of how above is done in `AsyncRequireAuthorizationLayer` in `tower-http` and I realized that I basically have to replicate its work.

I tried using it directly.  It was not easy to find out the underlying restrictions imposed by `AsyncRequireAuthorizationLayer`.  The example and test code do not use route integration and data types were not matching.

The key was in identifying that custom validator must return `UnsyncBoxBody<axum::body::Bytes, axum::Error>`.  I am not fully following why `UnsyncBoxBody` is used.

It all culminated in this work:  <https://github.com/kulak/axum-middleware/blob/main/src/jwt_auth_delegate.rs> and a corresponding test app <https://github.com/kulak/axum-middleware/blob/main/examples/jwt-auth-delegate/src/main.rs>. 

Hopefully, next one will be easier to write.

**Comment by davidpdrsn** [comment]
Did you read https://docs.rs/axum/0.6.20/axum/middleware/index.html#writing-middleware?

The recommendation is to use https://docs.rs/axum/0.6.20/axum/middleware/fn.from_fn.html instead of making a tower::Service directly. It’s way easier.

**Reply by kulak** [reply]
Yes, I read this and how AXUM was designed.

If I went with function, I would have application state a bit more disconnected.  I really wanted to accumulate validation configuration parameters in a way that was clearly connected.  That meant use of lower level abstractions.

Basically, validation layer has some configuration state.  Using application state to pass validation configuration to function, felt exposing private state across application where it is not necessary.  It just did not look right.

Otherwise, I would've went with functions.

**Reply by davidpdrsn** [reply]
> Using application state to pass validation configuration to function, felt exposing private state across application where it is not necessary

You can have a state type that you only pass to middleware::from_fn_with_state and don’t use anywhere else.

You can also use closure captures or Extension.

**Reply by sihayas** [reply]
Using application state to pass validation configuration to function, felt exposing private state across application where it is not necessary

I found the same issue personally. Regardless, the example in the docs is outdated it seems like.
